### DISCUSSION TITLE
How to deal with a package that only has tests? :)

### DISCUSSION THREAD

**Discussion by dpc** [ipetkov/crane#59] [Q&A] [answered]
I have a cargo package that has only tests. I'd like to plug it into things just for the `doChecks`.

```
> nix build .#minimint-tests
error: builder for '/nix/store/2s7yqxc15xyaq50y5a9qkak6dp338dcg-minimint-tests-0.0.1.drv' failed with exit code 101;
       last 10 log lines:
       > configuring
       > will append /build/source/.cargo-home/config.toml with contents of /nix/store/36g1bq0n87s6idcpks7pdp4daz9725w2-vendor-cargo-deps/config.toml
       > no configure script, doing nothing
       > setting CARGO_BUILD_RUSTFLAGS to "--remap-path-prefix /nix/store/36g1bq0n87s6idcpks7pdp4daz9725w2-vendor-cargo-deps=/sources"
       > building
       > cargo 1.62.1 (a748cf5a3 2022-06-08)
       > running: cargoBuildLog=$(mktemp cargoBuildLogXXXX.json)
       > cargo build --workspace --release --message-format json-render-diagnostics --bin minimint-tests >"$cargoBuildLog"
       >
       > error: no bin target named `minimint-tests`
```

I guess I'll just add an empty binary, but ideally this wouldn't be necessary.

**Comment by ipetkov** [comment, marked as answer]
> cargo build --workspace --release --message-format json-render-diagnostics --bin minimint-tests >"$cargoBuildLog"
>
> error: no bin target named `minimint-tests`

This looks like it might be a configuration error as the `minimint-tests` binary is explicitly being selected for building. Please double check whatever flags/arguments are being inherited in the build expression!

**Reply by dpc** [reply]
That's probably the case, but if I don't select it how do I trigger the tests in it? It's a leaf package. Do I just have to select the workspace for tests?

**Reply by dpc** [reply]
Oh, I should have probably add that this crate has only *integration* tests. There's no `./src/`. There's only `./tests/` :)

**Reply by ipetkov** [reply]
Can you show me the whole expression used to build it? I'm guessing something somewhere is setting `cargoExtraArgs = "--bin minimint-tests"`. Is it supposed to be `cargoExtraArgs = "-p minimint-tests"` ?

Using `--bin` tells cargo to build a specific binary, while using `-p` tells cargo to build only a specific _crate_

**Reply by dpc** [reply]
I've been introducing crane in minimint.

https://github.com/fedimint/minimint/blob/aeb36ce5cf6dbe5e660b3f460b5b467300ce2c76/flake.nix#L205

I can't remember which thing exactly where failing though (as there is bunch of existing `nix-shell` based things, that I just forwarded from `shell.nix` to `flake.nix` using `flake-compat`.

I workarounded it by just adding `[lib]`. https://github.com/fedimint/minimint/blob/aeb36ce5cf6dbe5e660b3f460b5b467300ce2c76/integrationtests/Cargo.toml#L6

BTW. I was trying to get a lot of stuff done at once, and didn't understand all the details, so if you think all of this should just work, maybe something else is at fault. I will probably dig a bit more into it in the future.

I guess it comes down to double checking how am I supposed to handle such a crate. I doesn't produce any binaries or libraries, but I want to run it as a test. Once you tell me how exactly to do it right, I'll try to do it, and come back to you with how it went. :)

**Reply by ipetkov** [reply]
The problem is right here: https://github.com/fedimint/minimint/blob/aeb36ce5cf6dbe5e660b3f460b5b467300ce2c76/flake.nix#L93

The flake is pretty complex and I don't have a good way to suggest how to fix it, but if the `pkg` function is always called with the _name of a crate_ and not the name of an individual binary output, you can change line 93 to use `cargoExtraArgs = "-p ${name}";` and it might work.

> I guess it comes down to double checking how am I supposed to handle such a crate. I doesn't produce any binaries or libraries, but I want to run it as a test. Once you tell me how exactly to do it right, I'll try to do it, and come back to you with how it went. :)

it should just work :)
I explicitly added a test case for it in #61 to make sure it keeps working too :+1:

**Reply by ipetkov** [reply]
@dpc here's a quick-n-dirty fix for your flake:

```diff
diff --git a/flake.nix b/flake.nix
index 2171943a..015ade91 100644
--- a/flake.nix
+++ b/flake.nix
@@ -83,16 +83,17 @@
         # all the dependencies (as dir) to help limit the
         # amount of things that need to rebuild when some
         # file change
-        pkg = { name, dir, extraDirs ? [ ] }: rec {
+        pkg = { name ? null , dir, extraDirs ? [ ] }: rec {
           package = craneLib.buildPackage (commonArgs // {
             cargoArtifacts = workspaceDeps;
-            pname = name;
 
             src = filterModules ([ dir ] ++ extraDirs) ./.;
 
-            cargoExtraArgs = "--bin ${name}";
             # if needed we will check the whole workspace at once with `workspaceAll`
             doCheck = false;
+          } // lib.optionalAttrs (name != null) {
+            pname = name;
+            cargoExtraArgs = "--bin ${name}";
           });
         };
 
@@ -180,7 +181,6 @@
         };
 
         minimint-tests = pkg {
-          name = "minimint-tests";
           dir = "integrationtests";
           extraDirs = [
             "client/cli"
diff --git a/integrationtests/Cargo.toml b/integrationtests/Cargo.toml
index 17531923..9cbeadc9 100644
--- a/integrationtests/Cargo.toml
+++ b/integrationtests/Cargo.toml
@@ -3,10 +3,6 @@ name = "minimint-tests"
 version = "0.1.0"
 edition = "2018"
 
-[lib]
-name = "minimint_tests"
-path = "src/main.rs"
-
 [[test]]
 name = "minimint-tests"
 path = "tests/tests.rs"
diff --git a/integrationtests/src/main.rs b/integrationtests/src/main.rs
deleted file mode 100644
index 35e54892..00000000
--- a/integrationtests/src/main.rs
+++ /dev/null
@@ -1,3 +0,0 @@
-pub fn main() {
-    // just here so the building system has a build target in this package
-}
```

**Reply by dpc** [reply]
Will try and report back. Thanks a lot!

**Reply by dpc** [reply]
Worked perfectly.
